### PR TITLE
update software installation script for EESSI 2021.02 pilot

### LINKED ISSUE #70: update build script for EasyBuild v4.3.3
Build script should be updated to use EasyBuild v4.3.3 once that has been released (ETA: week of Feb 22 '21)

### LINKED ISSUE #71: also use {linux,macos} subdirectory in software layer
cfr. compat layer

We need to discriminate between software installations on top of compat/linux/x86_86 vs compat/macos/x86_64, etc.

### PR DESCRIPTION
**Pull request by boegel**
Overview of changes:

* updates for EasyBuild v4.3.3 (fixes #70)
* also use `{linux,macos}` subdirectory in `2021.02/software/` (fixes #71)
* change check for Prefix environment (relying on start of `$PATH` value is problematic, sometimes additional paths are included at front)
* fix path to Lmod init script (changes due to https://github.com/EESSI/gentoo-overlay/pull/9?)
* filter additional dependencies (to use what's already provided in compat layer)
  * *not* filtering `Bison`, due to problems with `Qt5` using `yacc` from compat layer (see #49)
* stop using EasyBuild bootstrap script, use `pip install` + `eb --install-latest-eb-release`

(WIP, still testing it...)

### CODE REVIEW

**Comment by ocaisa** [comment]
We need to add an explicit build of UCX with the the option
```
configopts = '--enable-optimizations --enable-cma --enable-mt --with-verbs --with-rdmacm=$EPREFIX/usr --with-sysroot=$EPREFIX'
```

**Comment by boegel** [comment]
@ocaisa Is that to fix a particular problem? Can we fix it in a generic way on the EasyBuild side?

**Comment by ocaisa** [comment]
See https://github.com/EESSI/software-layer/issues/63 , I'm not sure it is a fit on the EB side since it is relying on you having certain system packages installed. For the record, the JSC configuration for UCX has way more options:
```
configopts = '--with-verbs '  # Build OpenFabrics support
configopts += '--without-java '
configopts += '--disable-doxygen-doc '

configopts += '--enable-optimizations '  # Enable machine-specific optimizations, default: NO
# configopts += '--enable-tuning '       # Enable parameter tuning in run-time, default: NO
configopts += '--enable-mt '           # Enable thread support in UCP and UCT, default: NO
configopts += '--disable-debug '
configopts += '--disable-logging '
configopts += '--disable-assertions '
configopts += '--disable-params-check '
configopts += '--disable-dependency-tracking '
configopts += '--with-cuda=$EBROOTCUDA '
configopts += '--enable-cma '           # Enable Cross Memory Attach
configopts += '--with-rc '              # Compile with IB Reliable Connection support
configopts += '--with-ud '              # Compile with IB Unreliable Datagram support
configopts += '--with-dc '              # Compile with IB Dynamic Connection support
configopts += '--with-mlx5-dv '         # Compile with mlx5 Direct Verbs support
configopts += '--with-ib-hw-tm '        # Compile with IB Tag Matching support
configopts += '--with-dm '              # Compile with Device Memory support
configopts += '--without-cm '           # Disable IB CM
configopts += '--with-avx '             # Compile with AVX
configopts += '--with-gdrcopy '         # Compile with GDRCopy
```

**Comment by ocaisa** [comment]
And the ComputeCanada customisation is at https://github.com/ComputeCanada/easybuild-computecanada-config/blob/b40236c4509d02ed7066ee91f6d3f298d500eba0/cc_hooks_gentoo.py#L533

(also their customisation for OpenMPI is pretty big, see https://github.com/ComputeCanada/easybuild-computecanada-config/blob/b40236c4509d02ed7066ee91f6d3f298d500eba0/cc_hooks_gentoo.py#L490)

File: EESSI-pilot-install-software.sh [hunk 1]
```diff
@@ -70,16 +70,15 @@ if [[ "$1" == "--generic" || "$EASYBUILD_OPTARCH" == "GENERIC" ]]; then
...
+# make sure we're in Prefix environment by which path to 'bash' command
+if [[ $(which bash) = ${EPREFIX}/bin/bash ]]; then
     echo_green ">> It looks like we're in a Gentoo Prefix environment, good!"
 else
     error "Not running in Gentoo Prefix environment, run '${EPREFIX}/startprefix' first!"
 fi
 
 echo ">> Initializing Lmod..."
-source $EPREFIX/usr/lmod/lmod/init/bash
+source $EPREFIX/usr/share/Lmod/init/bash
```
Comment (ocaisa) [inline EESSI-pilot-install-software.sh:81]: This update is required for bash initialisation as well (also for Magic Castle), will I open a separate PR for that?

Comment (boegel) [reply EESSI-pilot-install-software.sh:81]: Yes, let's do init script for `2021.02` in a separate PR.

**Comment by boegel** [comment]
@bedroge @ocaisa This should be good to go in now imho, so others can just run `./EESSI-pilot-install-software.sh` from a clone of this repo to install the `2021.02` pilot software layer...

**Comment by bedroge** [comment]
I was just going through the [project dashboard](https://github.com/orgs/EESSI/projects/4) to check if everything that we wanted to do for this version has been included, and saw this issue under software layer:
https://github.com/EESSI/software-layer/issues/67
I don't see this in the installation script though, should we still add this?

**Review by ocaisa — APPROVED** [review]

**Comment by ocaisa** [comment]
I wanted to make a PR to add RStudio Server and VS Code to the list as well, I can include ReFrame

**Comment by ocaisa** [comment]
PR for additional packages opened in #76

**Pull request merged**